### PR TITLE
ARROW-13553: [Doc] Add guidelines for code reviews

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@jorisvandenbossche @lidavidm @bkietz @jonkeane Feedback welcome on this.

Also spotting clumsy or incorrect wording would be very useful @thisisnic .

File: docs/source/developers/reviewing.rst [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+Guidelines
+==========
+
+Meta
+----
+
+**Use your common sense**.  These guidelines are not hard rules.
+Committers are expected to have sufficient expertise on their work
+areas to be able to modulate concerns if necessary.
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:45]: "modulate" isn't quite right here, but I'm not sure on the precise point intended so not sure what to suggest in its place, sorry

File: docs/source/developers/reviewing.rst [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+Finally, **these guidelines are not currently exhaustive**.
+
+Public API design
+-----------------
+
+- Public APIs should nudge users towards the most desirable constructs.
+  In other words, if there is a "best" way to do something, it should
+  ideally also be the most easily discoverable and the most concise to type.
+  For example, safe APIs should ideally be featured more prominently than
+  unsafe APIs that may crash or silently produce erronenous results on
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:59]:
```suggestion
  unsafe APIs that may crash or silently produce erroneous results on
```

File: docs/source/developers/reviewing.rst [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+  would be understandable without having to read the API docs.
+  Vague naming should be avoided; inaccurate naming is even worse as it
+  can mislead the reader and lead to buggy user code.
+
+- Be mindful of terminology.  Every project has (explicitly or tacitly) set
+  conventions about how to name important concepts; steering away from those
+  conventions increases the cognitive workload both for contributors and
+  users of the project.  Conversely, reusing a well-known term for a different
+  purpose than usual can also increase the cognitive workload and make
+  developers' life generally more difficult.
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:77]:
```suggestion
  developers' lives generally more difficult.
```

File: docs/source/developers/reviewing.rst [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+- If it is unsure whether an API is the right one for the task at hand,
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:79]:
```suggestion
- If you are unsure whether an API is the right one for the task at hand,
```

**Comment by bkmgit** [comment]
Would suggest to use precise descriptions but keep English language requirements as simple as possible.

File: docs/source/developers/reviewing.rst [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+  estimate is too imprecise) may be more critical than trying to improve
+  the common case by a small amount.
+
+Documentation
+-------------
+
+These guidelines should ideally apply to both prose documentation and
+in-code docstrings.
+
+- Look for ambiguous / sub-informative wording.  For example, "it is an error
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:140]:
```suggestion
- Look for ambiguous / uninformative wording.  For example, "it is an error
```

File: docs/source/developers/reviewing.rst [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+  if ..." is less informative than either "An error is raised if ... " or
+  "Behaviour is undefined if ..." (the first phrasing doesn't tell the
+  reader what actually *happens* on such an error).
+
+- Be mindful about spelling, grammar, expression, concision.  This may seem
+  like an obvious concern, but not all people are not accustomed to care about
+  this.
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:147]: Perhaps, instead of saying it's obvious but that not all people care about it, we could phrase it more positively, for example, saying that these things help smooth communication, or something like that?

File: docs/source/developers/reviewing.rst [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+* Some contributors are looking for a quick fix to a specific problem and
+  don't want to spend too much time on it.  Others on the contrary are eager
+  to learn and improve their contribution to make it conform to the
+  project's standards.  The latter kind of contributors are especially
+  valuable as they may become long-term contributors or even committers
+  to the project; it can therefore be a good strategy to prioritize
+  interactions with such contributors.
+
+* Some contributors may respond "I will fix it later, can we merge anyway?"
+  when a problem is pointed.  Unfortunately, whether the fix is really
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:215]:
```suggestion
  when a problem is pointed out to them.  Unfortunately, whether the fix will really be
```

File: docs/source/developers/reviewing.rst [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+  contributed soon in later PR is difficult to predict or enforce.  If
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:216]:
```suggestion
  contributed soon in a later PR, is difficult to predict or enforce.  If
```

File: docs/source/developers/reviewing.rst [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+  the contributor has shown to be trustable, it may be acceptable to do
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:217]:
```suggestion
  the contributor has previously demonstrated that they are reliable, it may be acceptable to do
```

File: docs/source/developers/reviewing.rst [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+=======================
+Reviewing contributions
+=======================
+
+Principles
+==========
+
+Arrow is a foundational project that will need to evolve over many years
+or even decades, while serving potentially millions of users.  We believe
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:26]:
```suggestion
or even decades, while serving potentially millions of users.  We believe that
```

File: docs/source/developers/reviewing.rst [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+
+Meta
+----
+
+**Use your common sense**.  These guidelines are not hard rules.
+Committers are expected to have sufficient expertise on their work
+areas to be able to modulate concerns if necessary.
+
+These guidelines are not listed in a particular order.  They are
+not intended to be used as a checklist.
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:48]:
```suggestion
These guidelines are not listed in a particular order or not intended to be used as a checklist.
```

**Review by thisisnic — COMMENTED** [review]
Looks good so far - various suggestions; some corrections and some just my own preferences so feel free to keep or disregard whatever.

Really like the section on social aspects - I think this is something that's really important to talk about explicitly so people know what to expect :)

**Comment by pitrou** [comment]
> Would suggest to use precise descriptions but keep English language requirements as simple as possible.

What do you mean by that? Is the proposed document too difficult to read?
(FTR, English is not my native language)

File: docs/source/developers/reviewing.rst [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+  any progress, it is also possible to take it up.  Out of politeness,
+  it is however better to ask the contributor first.
+
+* Some contributors are looking for a quick fix to a specific problem and
+  don't want to spend too much time on it.  Others on the contrary are eager
+  to learn and improve their contribution to make it conform to the
+  project's standards.  The latter kind of contributors are especially
+  valuable as they may become long-term contributors or even committers
+  to the project; it can therefore be a good strategy to prioritize
+  interactions with such contributors.
```
Comment (lidavidm) [inline docs/source/developers/reviewing.rst:212]: I get the point here, but I worry this sounds a little exclusive; perhaps we can say instead that we should help people who are interested in getting further involved with the community to do so?

File: docs/source/developers/reviewing.rst [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+==========
+
+Arrow is a foundational project that will need to evolve over many years
+or even decades, while serving potentially millions of users.  We believe
+being meticulous when reviewing brings greater rewards to the project than
+being lenient and aiming for quick merges.
+
+Fixing potential issues after a Pull Request (PR) merge is more costly since
+it forces the developer to context-switch back to work that was thought
+to be finished.  Moreover, an late-minute API change may propagate to
```
[the remark this answers is not in the document]
Comment (jonkeane) [reply docs/source/developers/reviewing.rst:32]: "last minute" is more idiomatic for me too

File: docs/source/developers/reviewing.rst [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+  is of course very subjective, but two weeks can be a reasonable
+  heuristic).  If you cannot allocate time soon, do say it explicitly.
+  If you don't have the answer to a question, do say it explicitly.
+  Saying "I don't have time immediately but I will come back later,
+  feel free to ping if I seem to have forgotten" or "Sorry, I am out of depth
+  here" is always better than saying nothing and letting the interlocutor
+  wondering.
+
+* If you know someone who may help on a blocking issue and past experience
+  suggests they can be available for that, feel free to add them to the
```
Comment (jonkeane) [inline docs/source/developers/reviewing.rst:193]: "available" here seems orthogonal to expertise shown through past experience. Should we phrase this about both expertise and that the ping is a way to assess availability?

Comment (jonkeane) [reply docs/source/developers/reviewing.rst:193]: I know I am frequently hesitant to `@` someone, but I should more frequently and they can say either "sorry I don't have time" or "oh this is the answer..."

File: docs/source/developers/reviewing.rst [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+  project's standards.  The latter kind of contributors are especially
+  valuable as they may become long-term contributors or even committers
+  to the project; it can therefore be a good strategy to prioritize
+  interactions with such contributors.
+
+* Some contributors may respond "I will fix it later, can we merge anyway?"
+  when a problem is pointed.  Unfortunately, whether the fix is really
+  contributed soon in later PR is difficult to predict or enforce.  If
+  the contributor has shown to be trustable, it may be acceptable to do
+  as suggested.  Otherwise, it is better to decline the suggestion.
```
Comment (jonkeane) [inline docs/source/developers/reviewing.rst:218]: I wonder if framing this in terms of in-scope/out-of-scope decisions and follow-on Jiras. Something along the lines of: 

Our general policy is to not introduce regressions or merge PRs that require follow ons to function correctly (though exceptions to this can be made). What changes are in-scope for a PR and what changes might/could/should be pushed out of scope and have a Jira created should be determined in collaboration between the authors and the reviewers

File: docs/source/developers/reviewing.rst [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+
+* If a PR is generally ready for merge apart from trivial or uncontroversial
+  concerns, the reviewer may decide to push changes themselves to the
+  PR instead of asking the contributor to make the changes.
```
Comment (jonkeane) [inline docs/source/developers/reviewing.rst:222]: Do we want a note about that same "it's best to ask for permission" or possibly we could frame this as "please don't take offense if this happens, it's common and is not intended to be a slight"

**Review by jonkeane — COMMENTED** [review]
This is really great, thanks for everything you've got here already.

Would it be a good idea to also mention + link to our code of conduct and call out (or reiterate) that reviews (both from the reviewer and the author) must respect the code of conduct (I don't think we've had any issues on this front, but it would be good to have a call out here to remind people).

I also wonder if it would be good to have a section about reviews being collaborative and everyone should try their best to keep them professional and solutions oriented? This cuts in both ways: it can be easy for a PR author to feel a bit down/attacked if there are _lots_ of changes requested (and mentioning this + framing the review process as a collaborative one might help resolve some of that) and it also reminds reviewers to some empathy in their reviewing.

**Comment by bkmgit** [comment]
> > Would suggest to use precise descriptions but keep English language requirements as simple as possible.
> 
> What do you mean by that? Is the proposed document too difficult to read? (FTR, English is not my native language)

Expect many contributors use languages other than English.  Making reviews easy to read would likely be helpful.  Perhaps this could be added in the social aspects section?

File: docs/source/developers/reviewing.rst [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+
+These guidelines should ideally apply to both prose documentation and
+in-code docstrings.
+
+- Look for ambiguous / sub-informative wording.  For example, "it is an error
+  if ..." is less informative than either "An error is raised if ... " or
+  "Behaviour is undefined if ..." (the first phrasing doesn't tell the
+  reader what actually *happens* on such an error).
+
+- Be mindful about spelling, grammar, expression, concision.  This may seem
```
Comment (bkmgit) [inline docs/source/developers/reviewing.rst:145]: Maybe 
```suggestion
- Be concise. Use correct spelling and grammar. 
```

File: docs/source/developers/reviewing.rst (hunk 6 above)
Comment (bkmgit) [reply docs/source/developers/reviewing.rst:147]: Yes, possibly something like "Clear communication in a group with a range of linguistic and programming backgrounds enables more efficient collaborative development and maintenance."

File: docs/source/developers/reviewing.rst [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+Arrow is a foundational project that will need to evolve over many years
+or even decades, while serving potentially millions of users.  We believe
+being meticulous when reviewing brings greater rewards to the project than
+being lenient and aiming for quick merges.
+
+Fixing potential issues after a Pull Request (PR) merge is more costly since
+it forces the developer to context-switch back to work that was thought
+to be finished.  Moreover, an late-minute API change may propagate to
+other parts of the project and require more attention than if done up front.
+And there are social issues with asking a volunteer to go back to work
```
[the remark this answers is not in the document]
Comment (jorisvandenbossche) [reply docs/source/developers/reviewing.rst:34]: I fully agree with this paragraph, but at the same time it can also be "demotivating" if a PR drags on for a very long time (although that can have many reasons of course, not necessarily due to the review). So maybe we could just says _something_ about that we still try to make contributing a rewarding process or something like that.

**Comment by jorisvandenbossche** [comment]
> Would it be a good idea to also mention + link to our code of conduct and call out (or reiterate) that reviews (both from the reviewer and the author) must respect the code of conduct (

+1

> I also wonder if it would be good to have a section about reviews being collaborative and everyone should try their best to keep them professional and solutions oriented? This cuts in both ways: it can be easy for a PR author to feel a bit down/attacked if there are lots of changes requested (and mentioning this + framing the review process as a collaborative one might help resolve some of that) and it also reminds reviewers to some empathy in their reviewing.

Yes, I agree it would be nice to have a section on this, explicitly calling this out. 
Maybe we could also mention a general guideline along the lines of "assume good intentions" (eg if you might feel attacked by the review), as well as avoiding "strong" wording on the reviewers side (like "this is wrong" can also be phrased as "I think we should do it this way instead"). This blogpost of Gael Varoqaux is a nice reference about this IMO: http://gael-varoquaux.info/programming/technical-discussions-are-hard-a-few-tips.html

File: docs/source/developers/reviewing.rst [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+on something that was accepted as finished.
+
+Guidelines
+==========
+
+Meta
+----
+
+**Use your common sense**.  These guidelines are not hard rules.
```
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:43]: Would "use your own judgement" be better?

File: docs/source/developers/reviewing.rst (hunk 1 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:45]: Hmm... "calibrate" perhaps? "adjust"?

File: docs/source/developers/reviewing.rst (hunk 5 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:140]: I'll use "poorly informative".

File: docs/source/developers/reviewing.rst (hunk 6 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:147]: Hmm, how something like:
> People have varying degrees of adherence to these concerns, but they contribute to a better documentation.

File: docs/source/developers/reviewing.rst [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+Robustness
+----------
+
+- Arrow is a set of open source libraries and will be used in a very wide
+  array of contexts (including fiddling with deliberately artificial data
+  at a Jupyter interpreter prompt).  If you are writing a public API, make
+  sure that it won't crash or produce undefined behaviour on unusual (but
+  valid) inputs.
+
+- When a non-trivial algorithm is implemented, defense-in-depth checks can
```
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:94]: You're right, that should be "defensive coding".

File: docs/source/developers/reviewing.rst (hunk 15 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:218]: Ha, nice suggestion. I may even expand this a little.

File: docs/source/developers/reviewing.rst (hunk 1 above)
Comment (thisisnic) [reply docs/source/developers/reviewing.rst:45]: I think the bit where I'm confused is knowing what is being calibrated/adjusted, and who has concerns about what - "modulate concerns" doesn't quite work.

File: docs/source/developers/reviewing.rst (hunk 6 above)
Comment (thisisnic) [reply docs/source/developers/reviewing.rst:147]: What about if, instead of talking about when people don't do the thing we want them to do, we phrase it in terms of the result of them doing the thing we want them to do:

>  Clear communication is essential for effective collaboration with people from a wide range of backgrounds, and contributes to better documentation.

File: docs/source/developers/reviewing.rst [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+
+* Reviewing is a communication between the contributor and the reviewer.
+  Avoid letting questions or comments unanswered for too long ("too long"
+  is of course very subjective, but two weeks can be a reasonable
+  heuristic).  If you cannot allocate time soon, do say it explicitly.
+  If you don't have the answer to a question, do say it explicitly.
+  Saying "I don't have time immediately but I will come back later,
+  feel free to ping if I seem to have forgotten" or "Sorry, I am out of depth
+  here" is always better than saying nothing and letting the interlocutor
+  wondering.
```
[the remark this answers is not in the document]
Comment (thisisnic) [reply docs/source/developers/reviewing.rst:190]: I am going to admit, as someone whose first language is English, I thought it had negative connotations until I looked it up just now and found out it didn't - I think I mixed it up with "interloper"!

File: docs/source/developers/reviewing.rst (hunk 6 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:147]: Ok :-)

File: docs/source/developers/reviewing.rst (hunk 12 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:212]: Hmm, I may simply remove the last proposition ("it can therefore be a good strategy...").

File: docs/source/developers/reviewing.rst [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+Reviewing contributions
+=======================
+
+Principles
+==========
+
+Arrow is a foundational project that will need to evolve over many years
+or even decades, while serving potentially millions of users.  We believe
+being meticulous when reviewing brings greater rewards to the project than
+being lenient and aiming for quick merges.
```
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:28]: I'll add something to that tune, thanks.

File: docs/source/developers/reviewing.rst [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+- Think about performance, but do not obsess over it.  Algorithmic complexity
+  is important if input size may be "large" (the meaning of large depends
+  on the context: use your own expertise to decide!).  Micro-optimizations
+  improving performance by 20% or more on performance-sensitive functionality
+  may be useful as well; lesser micro-optimizations may not be worth the
+  time spent on them, especially if they lead to more complicated code.
+
+- If performance is important, measure it.  Do not satisfy yourself with
+  guesses and intuitions (which may be founded on incorrect assumptions
+  about the compiler or the hardware).
```
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:121]: I need to add links, yes.

File: docs/source/developers/reviewing.rst [hunk 24]
```diff
@@ -0,0 +1,226 @@
...
+
+- Try to avoid trying to trick the compiler/interpreter/runtime by writing
+  the code in a certain way, unless it's really important.  These tricks
+  are generally dependent on platform details that may become obsolete,
+  they are often brittle, and they can make code harder to maintain
+  (a common question that can block contributors is "what should I do
+  about this weird hack that my changes need to remove"?).
+
+- Avoiding worst-case degenerescence (such as memory blowups when a size
```
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:130]: It might be a term of art in French only, sorry :-(

[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:130]: Perhaps "degradation" works?

[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:130]: Or "catastrophic worst-case degradation"?

[the remark this answers is not in the document]
Comment (lidavidm) [reply docs/source/developers/reviewing.rst:130]: I like the second one, or perhaps just "Avoiding degenerate behavior…" in general

File: docs/source/developers/reviewing.rst (hunk 21 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:190]: Ok, let's remove the interloper :-)

File: docs/source/developers/reviewing.rst (hunk 14 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:193]: Hmm, who about:

"If you know someone who has the competence to help on a blocking issue and past experience suggests they may be willing to do so, feel free to add them to the discussion (for example by gently pinging their Github handle)."

File: docs/source/developers/reviewing.rst (hunk 16 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:222]: I think the proposed phrase is out of scope for this document, as it's intended for reviewers not submitters.

File: docs/source/developers/reviewing.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:130]: Ah, "degenerate behaviour" sounds fine.

[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:130]: I'll change to: "Avoiding rough edges or degenerate behaviour (such as memory blowups when a size estimate is inaccurately large)"

File: docs/source/developers/reviewing.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (pitrou) [reply docs/source/developers/reviewing.rst:34]: I am reworking back this part entirely based on @jonkeane's suggestion in another comment :-)

File: docs/source/python/csv.rst [hunk 25]
```diff
@@ -16,7 +16,7 @@
 .. under the License.
 
 .. currentmodule:: pyarrow.csv
-.. _csv:
+.. _py-csv:
```
Comment (pitrou) [inline docs/source/python/csv.rst:19]: Note @jorisvandenbossche : we (meaning I :-)) took the bad habit of not prefixing reference names with their language. This won't be a problem anymore if we split project docs, but I had to disambiguate here.

File: docs/source/developers/reviewing.rst [hunk 26]
```diff
@@ -0,0 +1,256 @@
...
+
+Principles
+==========
+
+Arrow is a foundational project that will need to evolve over many years
+or even decades, while serving potentially millions of users.  We believe
+that being meticulous when reviewing brings greater rewards to the project
+than being lenient and aiming for quick merges.
+
+Code review like this leads to better quality code, more people who are
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:30]:
```suggestion
Code reviews like this leads to better quality code, more people who are
```

File: docs/source/developers/reviewing.rst [hunk 27]
```diff
@@ -0,0 +1,256 @@
...
+engaged with and understand the code being changed, and a generally
+healthier project with more room to grow and accomodate emerging needs.
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:32]:
```suggestion
healthier project with more room to grow and accommodate emerging needs.
```

File: docs/source/developers/reviewing.rst [hunk 28]
```diff
@@ -0,0 +1,256 @@
...
+----
+
+**Use your own judgement**.  These guidelines are not hard rules.
+Committers are expected to have sufficient expertise on their work
+areas to be able to adjust their approach based on any concerns they have.
+
+These guidelines are not listed in a particular order or not intended to be
+used as a checklist.
+
+Finally, these guidelines are not currently exhaustive.
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:48]:
```suggestion
Finally, these guidelines are not exhaustive.
```
nit: can you ever have exhaustive guidelines on code reviews given, as mentioned above, the need for discretion?

File: docs/source/developers/reviewing.rst [hunk 29]
```diff
@@ -0,0 +1,256 @@
...
+  follow-ons to function correctly (though exceptions to this can be made).
+  Making necessary changes after a merge is more costly both for the
+  contributor and the reviewer, but also for other developers who may be
+  confused if they hit problems introduced by a merged PR.
+
+* What changes are in-scope for a PR and what changes might/could/should be
+  pushed out of scope and have a follow-up JIRA created should be determined
+  in collaboration between the authors and the reviewers.
+
+* When a large functionality is being contributed and it seems desirable
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:63]:
```suggestion
* When a large piece of functionality is being contributed and it seems desirable
```

File: docs/source/developers/reviewing.rst [hunk 30]
```diff
@@ -0,0 +1,256 @@
...
+These guidelines should ideally apply to both prose documentation and
+in-code docstrings.
+
+* Look for ambiguous / poorly informative wording.  For example, *"it is an
+  error if ..."* is less informative than either *"An error is raised if ... "*
+  or *"Behaviour is undefined if ..."* (the first phrasing doesn't tell the
+  reader what actually *happens* on such an error).
+
+* When reviewing documentation changes (or prose snippets, in general),
+  be mindful about spelling, grammar, expression, concision.  Clear
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:166]:
```suggestion
  be mindful about spelling, grammar, expression, and concision.  Clear
```

File: docs/source/developers/reviewing.rst [hunk 31]
```diff
@@ -0,0 +1,256 @@
...
+  communication is essential for effective collaboration with people
+  from a wide range of backgrounds, and contributes to better documentation.
+
+* Some contributors do not have English as a native language (and perhaps
+  neither do you).  It is advised to help them and/or ask for external help
+  if needed.
+
+* Cross-linking increases the global value of documentation.  Sphinx especially
+  has great cross-linking capabilities (including topic references, glossary
+  terms, API references), be sure to make use of them!
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:176]:
```suggestion
  terms, API references), so be sure to make use of them!
```

File: docs/source/developers/reviewing.rst [hunk 32]
```diff
@@ -0,0 +1,256 @@
...
+  worthwhile to fine-tune testing parameters to balance the usefulness
+  of tests against the cost of running them (especially where stress tests
+  are involved, since they tend to imply execution over large datasets).
+
+
+Social aspects
+==============
+
+* Reviewing is a communication between the contributor and the reviewer.
+  Avoid letting questions or comments unanswered for too long ("too long"
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:208]:
```suggestion
  Avoid letting questions or comments remain unanswered for too long ("too long"
```

File: docs/source/developers/reviewing.rst [hunk 33]
```diff
@@ -0,0 +1,256 @@
...
+  my depth here"* is always better than saying nothing and leaving the
+  other person wondering.
+
+* If you know someone who has the competence to help on a blocking issue
+  and past experience suggests they may be willing to do so, feel free to
+  add them to the discussion (for example by gently pinging their Github
+  handle).
+
+* If the contributor has stopped giving feedback or updating their PR,
+  perhaps they're not interested anymore, but perhaps also they're stuck
```
Comment (thisisnic) [inline docs/source/developers/reviewing.rst:223]:
```suggestion
  perhaps they're not interested any more, but perhaps also they're stuck
```

**Review by thisisnic — COMMENTED** [review]
Other than a few minor word insertions/deletions, this is looking great!

File: docs/source/developers/reviewing.rst (hunk 28 above)
Comment (pitrou) [reply docs/source/developers/reviewing.rst:48]: Of course :-D

File: docs/source/developers/reviewing.rst (hunk 16 above)
Comment (westonpace) [reply docs/source/developers/reviewing.rst:249]: No change needed but I just wanted to mention that I would probably suggest this document to newer submitter as well.  Much of the content here is helpful for understanding why the code is written the way it is.  It would also help with common questions new contributors might face (e.g. how much should I be testing my code?)
